packages/api/src/rich-text/detection.ts

```diff
@@ -0,0 +1,79 @@
+import TLDs from 'tlds'
+import { AppBskyRichtextFacet } from '../client'
+import { UnicodeString } from './unicode'
+
+export type Facet = AppBskyRichtextFacet.Main
+
+export function detectFacets(text: UnicodeString): Facet[] | undefined {
+  let match
+  const facets: Facet[] = []
+  {
+    // mentions
+    const re = /(^|\s|\()(@)([a-zA-Z0-9.-]+)(\b)/g
+    while ((match = re.exec(text.utf16))) {
+      if (!isValidDomain(match[3]) && !match[3].endsWith('.test')) {
+        continue // probably not a handle
+      }
+
+      const start = text.utf16.indexOf(match[3], match.index) - 1
+      facets.push({
+        $type: 'app.bsky.richtext.facet',
```

Comment:
I can't think of anywhere the `$type` is strictly necessary on facets yet (is there one perhaps I've forgotten?), but either way I see no issue including it.